File: parser/src/main/java/io/ticofab/androidgpxparser/parser/domain/Route.java
package io.ticofab.androidgpxparser.parser.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Route {
    private final List<RoutePoint> mRoutePoints;
    private final String mRouteName;
    private final String mRouteDesc;
    private final String mRouteCmt;
    private final String mRouteSrc;
    private final Integer mRouteNumber;
    private final Link mRouteLink;
    private final String mRouteType;

    private Route(Builder builder) {
        mRoutePoints = Collections.unmodifiableList(new ArrayList<>(builder.mRoutePoints));
        mRouteName = builder.mRouteName;
        mRouteDesc = builder.mRouteDesc;
        mRouteCmt = builder.mRouteCmt;
        mRouteSrc = builder.mRouteSrc;
        mRouteNumber = builder.mRouteNumber;
        mRouteLink = builder.mRouteLink;
        mRouteType = builder.mRouteType;
    }

    public List<RoutePoint> getRoutePoints() {
        return mRoutePoints;
    }

    public String getRouteName() {
        return mRouteName;
    }

    public String getRouteDesc() {
        return mRouteDesc;
    }

    public String getRouteCmt() {
        return mRouteCmt;
    }

    public String getRouteSrc() {
        return mRouteSrc;
    }

    public Integer getRouteNumber() {
        return mRouteNumber;
    }

    public Link getRouteLink() {
        return mRouteLink;
    }

    public String getRouteType() {
        return mRouteType;
    }

    public static class Builder {
        private List<RoutePoint> mRoutePoints;
        private String mRouteName;
        private String mRouteDesc;
        private String mRouteCmt;
        private String mRouteSrc;
        private Integer mRouteNumber;
        private Link mRouteLink;
        private String mRouteType;

        public Builder setRoutePoints(List<RoutePoint> routePoints) {
            mRoutePoints = routePoints;
            return this;
        }

        public Builder setRouteName(String routeName) {
            mRouteName = routeName;
            return this;
        }

        public Builder setRouteDesc(String routeDesc) {
            mRouteDesc = routeDesc;
            return this;
        }

        public Builder setRouteCmt(String routeCmt) {
            mRouteCmt = routeCmt;
            return this;
        }

        public Builder setRouteSrc(String routeSrc) {
            mRouteSrc = routeSrc;
            return this;
        }

        public Builder setRouteNumber(Integer routeNumber) {
            mRouteNumber = routeNumber;
            return this;
        }

        public Builder setRouteLink(Link routeLink) {
            mRouteLink = routeLink;
            return this;
        }

        public Builder setRouteType(String routeType) {
            mRouteType = routeType;
            return this;
        }

        public Route build() {
            return new Route(this);
        }
    }
}

File: parser/src/main/kotlin/io/ticofab/androidgpxparser/parser/domain/Route.kt
package io.ticofab.androidgpxparser.parser.domain

class Route private constructor(builder: Builder) {
    val routePoints = builder.routePoints?.toList()
    val routeName = builder.routeName
    val routeDesc = builder.routeDesc
    val routeCmt = builder.routeCmt
    val routeSrc = builder.routeSrc
    val routeNumber = builder.routeNumber
    val routeLink = builder.routeLink
    val routeType = builder.routeType

    class Builder {
        var routePoints: List<RoutePoint>? = null
            private set
        var routeName: String? = null
            private set
        var routeDesc: String? = null
            private set
        var routeCmt: String? = null
            private set
        var routeSrc: String? = null
            private set
        var routeNumber: Int? = null
            private set
        var routeLink: Link? = null
            private set
        var routeType: String? = null
            private set

        fun setRoutePoints(routePoints: List<RoutePoint>?) = apply { this.routePoints = routePoints }

        fun setRouteName(routeName: String?) = apply { this.routeName = routeName }

        fun setRouteDesc(routeDesc: String?) = apply { this.routeDesc = routeDesc }

        fun setRouteCmt(routeCmt: String?) = apply { this.routeCmt = routeCmt }

        fun setRouteSrc(routeSrc: String?) = apply { this.routeSrc = routeSrc }

        fun setRouteNumber(routeNumber: Int?) = apply { this.routeNumber = routeNumber }

        fun setRouteLink(routeLink: Link?) = apply { this.routeLink = routeLink }

        fun setRouteType(routeType: String?) = apply { this.routeType = routeType }

        fun build() = Route(this)
    }
}